### ISSUE TITLE
[BUG] Opensearch Bulk API doesn't support bytes even if Serializer typing says it does

### ISSUE THREAD

**Issue by roshanjrajan-zip** [opensearch-project/opensearch-py#488]
### What is the bug?
I wanted to use an alternative serializer to speed up performance. Looking at the original JsonSerializer, if the type is a [`string_types` which is either str or bytes](https://github.com/opensearch-project/opensearch-py/blob/c8b04a5246dad6dcec05cf6f5bbf2eb14a9ec398/opensearchpy/compat.py#L48), it will return the data without using the serializer. So assumably any serializer that returns str or bytes is also valid. However, the following serializer using orjson returns bytes and [bulk API only accepts strings](https://github.com/opensearch-project/opensearch-py/blob/c8b04a5246dad6dcec05cf6f5bbf2eb14a9ec398/opensearchpy/helpers/actions.py#L117C1-L122C51) as it tries and encode the data using utf8. The [inbuilt JSONSerializer ](https://github.com/opensearch-project/opensearch-py/blob/c8b04a5246dad6dcec05cf6f5bbf2eb14a9ec398/opensearchpy/serializer.py#L151C1-L153) says it takes string_types as well.
```python
class OrjsonSerializer(JSONSerializer):
    def dumps(self, data: Any):
        if isinstance(data, string_types):
            return data

        try:
        	# The .decode("utf-8") was added to convert bytes to string
            return orjson.dumps(data, default=self.default).decode("utf-8") 
        except Exception as exc:
            raise SerializationError(data, exc) from exc

    def loads(self, s: Any):
        try:
            return orjson.loads(s)
        except Exception as exc:
            raise SerializationError(s, exc) from exc
```

### How can one reproduce the bug?
Return bytes from the Json Serializer instead of a string.

### What is the expected behavior?
I'm not sure what the correct answer is but it should either only accepts strings or have a separate code path if the type is bytes

### What is your host/environment?
Debian 5.10.191-1

### Do you have any screenshots?
![image](https://github.com/opensearch-project/opensearch-py/assets/128418999/fab8fc81-fafd-4dfd-8546-5beea783de63)

### Do you have any additional context?
_Add any other context about the problem._

**Comment by saimedhi** [comment]
@roshanjrajan-zip, Thanks for spotting the issue. Your proposed solution is appreciated – please feel free to contribute!

**Comment by roshanjrajan-zip** [comment]
@saimedhi Which is the preferred answer here? To only return a str or to support bytes?

**Comment by saimedhi** [comment]
In my view, supporting both strings and bytes would be ideal solution. Adding a mechanism to handle bytes would be great.

**Comment by saimedhi** [comment]
@dblock, @wbeckler  Please take a look at this issue.

**Comment by dblock** [comment]
So `bulk` expects a line-delimited JSON and the rest of the APIs expect proper JSON. Both are strings, aren't they? Isn't the  custom serializer the issue, shouldn't it be producing strings in all cases (like the `.decode("utf-8")` you've added)? What would be a compelling reason to change the default to pass around raw bytes?

**Comment by roshanjrajan-zip** [comment]
I don't think there is necessary a compelling reason to support bytes, just that the code is confusing in that if bytes were passed to the serializer, then it would return bytes immediately without even using the serializer. The bug is more about the typing check being inconsistent with the use case in the bulk api. I think a good answer could be replacing the instance check with `isinstance(data, str)`.

**Comment by dblock** [comment]
Ok, I think I understand. I agree on `isinstance(data, str)`, want to give it a try?

**Comment by roshanjrajan-zip** [comment]
Yeah we can do this! But is `string_types` supposed to include bytes? Should it just be str?

**Comment by dblock** [comment]
I think you know as much as I (we) do! ;)

**Comment by roshanjrajan-zip** [comment]
Well it looks like the elasticsearch team supports bytes only for this - https://github.com/elastic/elastic-transport-python/blob/main/elastic_transport/_serializer.py

**Comment by dblock** [comment]
@roshanjrajan-zip we aren't following changes in ES clients, but since those are APLv2 please feel free to port that code in a PR here
